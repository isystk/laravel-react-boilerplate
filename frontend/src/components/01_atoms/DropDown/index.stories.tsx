import { Meta, Story } from '@storybook/react'
import React from 'react'
import DropDown, { DropDownProps } from './index'

export default {
  title: '01_atoms/DropDown',
  component: DropDown,
} as Meta

const Template: Story = () => {
  const props: DropDownProps = {
    label: 'サンプル太郎',
    items: [
      { label: 'リンク1', link: 'http://link1' },
      { label: 'リンク2', link: 'http://link2' },
      { label: 'リンク3', link: 'http://link3' },
    ],
  }
  return <DropDown {...props} />
}

export const Primary = Template.bind({})
Primary.storyName = 'プライマリ'
Primary.args = {}
