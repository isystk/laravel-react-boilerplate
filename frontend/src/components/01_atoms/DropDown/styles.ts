import { css, keyframes } from '@emotion/css'

export const dropDown = css``
