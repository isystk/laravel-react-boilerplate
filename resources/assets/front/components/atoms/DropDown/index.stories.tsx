import { useState } from 'react';
import DropDown from './index';
import type { Meta, StoryFn } from '@storybook/react';

export default {
  title: 'Components/Atoms/DropDown',
  component: DropDown,
  tags: ['autodocs'],
} as Meta<typeof DropDown>;

export const Default: StoryFn = () => {
  const [selected, setSelected] = useState<string | null>(null);

  const items = [
    { text: '選択肢1', onClick: () => setSelected('選択肢1') },
    { text: '選択肢2', onClick: () => setSelected('選択肢2') },
    { text: '選択肢3', onClick: () => setSelected('選択肢3') },
  ];

  return (
    <div>
      <DropDown text="メニューを開く" items={items} />
      {selected && <p>選択された項目: {selected}</p>}
    </div>
  );
};

export const NoActions: StoryFn = () => (
  <DropDown
    text="クリックして開く"
    items={[{ text: 'アクションなし1' }, { text: 'アクションなし2' }]}
  />
);
