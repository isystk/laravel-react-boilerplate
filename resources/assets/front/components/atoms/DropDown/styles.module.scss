
.dropdown {
    position: relative;
}

.dropdownToggle {
    color: #4b5563;
    cursor: pointer;
    outline: none;
    background: none;
    border: none;
}

.dropdownMenu {
    position: absolute;
    right: 0;
    margin-top: 4px;
    width: 160px;
    background-color: white;
    border: 1px solid #d1d5db;
    border-radius: 0.375rem;
    box-shadow: 0 1px 2px rgba(0, 0, 0, 0.05);
    z-index: 10;
}

.dropdownItem {
    display: block;
    width: 100%;
    text-align: left;
    padding: 8px 16px;
    background: none;
    border: none;
    cursor: pointer;
}

.dropdownItem:hover {
    background-color: #f3f4f6;
}
