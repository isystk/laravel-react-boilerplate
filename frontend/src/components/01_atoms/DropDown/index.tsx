import React, { FC, useState } from 'react'
import { ContainerProps } from 'types'
import { connect } from '@/components/hoc'
import * as styles from './styles'
import { useRouter } from 'next/router'

export type DropDownItem = {
  label: string
  link: string | (() => void)
}

/** DropDownProps Props */
export type DropDownProps = {
  label: string
  items: DropDownItem[]
}
/** Presenter Props */
export type PresenterProps = DropDownProps & {
  router
  showMenu
  setShowMenu
}

/** Presenter Component */
const DropDownPresenter: FC<PresenterProps> = ({
  label,
  items = [],
  router,
  showMenu,
  setShowMenu,
}) => (
  <>
    <div className="relative">
      <button
        id="dropdownNavbarLink"
        data-dropdown-toggle="dropdownNavbar"
        className={`${styles.dropDown} flex items-center justify-between md:py-2 pl-3 md:pr-4 rounded md:border-0 md:p-0 `}
        onClick={() => setShowMenu(!showMenu)}
      >
        <span className="w-32 md:w-48 md:pr-4 truncate">{label}</span>
        <svg
          className="w-5 h-5 ml-1"
          aria-hidden="true"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            fillRule="evenodd"
            d="M5.293 7.293a1 1 0 011.414 0L10 10.586l3.293-3.293a1 1 0 111.414 1.414l-4 4a1 1 0 01-1.414 0l-4-4a1 1 0 010-1.414z"
            clipRule="evenodd"
          ></path>
        </svg>
      </button>
      <div
        id="dropdownNavbar"
        className={`${
          showMenu ? '' : 'hidden'
        } z-10 font-normal bg-white divide-y divide-gray-100 rounded-lg shadow absolute right-0 w-48`}
      >
        <ul
          className="py-2 text-sm text-gray-700 dark:text-gray-400"
          aria-labelledby="dropdownLargeButton"
        >
          {items.map((item, idx) => (
            <li key={idx}>
              <a
                href="#"
                className="block px-4 py-2 hover:bg-gray-100 dark:hover:bg-gray-600 dark:hover:text-white"
                onClick={
                  typeof item.link === 'string'
                    ? (e) => {
                        e.preventDefault()
                        router.push(item.link)
                      }
                    : item.link
                }
              >
                {item.label}
              </a>
            </li>
          ))}
        </ul>
      </div>
    </div>
  </>
)

/** Container Component */
const DropDownContainer: React.FC<
  ContainerProps<DropDownProps, PresenterProps>
> = ({ presenter, ...props }) => {
  const router = useRouter()
  const [showMenu, setShowMenu] = useState(false)
  return presenter({
    router,
    showMenu,
    setShowMenu,
    ...props,
  })
}

export default connect<DropDownProps, PresenterProps>(
  'DropDown',
  DropDownPresenter,
  DropDownContainer
)
