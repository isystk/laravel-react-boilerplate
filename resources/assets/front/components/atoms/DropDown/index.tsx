import { useState, useEffect, useRef } from 'react';
import styles from './styles.module.scss';

type Props = {
  text: string;
  items: Array<{ text: string; onClick: () => void }>;
  className?: string;
};

const DropDown = ({ text, items, className }: Props) => {
  const [isOpen, setOpen] = useState(false);
  const dropdownRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (dropdownRef.current && !dropdownRef.current.contains(event.target as Node)) {
        setOpen(false);
      }
    };

    document.addEventListener('click', handleClickOutside);
    return () => {
      document.removeEventListener('click', handleClickOutside);
    };
  }, []);

  return (
    <div ref={dropdownRef} className={`${styles.dropdown} ${className}`}>
      <button
        className={`btn ${styles.dropdownToggle}`}
        onClick={() => {
          setOpen(!isOpen);
        }}
      >
        {text}
        <svg
          className="-mr-1 size-5 text-gray-400"
          viewBox="0 0 20 20"
          fill="currentColor"
          aria-hidden="true"
          data-slot="icon"
        >
          <path
            fillRule="evenodd"
            d="M5.22 8.22a.75.75 0 0 1 1.06 0L10 11.94l3.72-3.72a.75.75 0 1 1 1.06 1.06l-4.25 4.25a.75.75 0 0 1-1.06 0L5.22 9.28a.75.75 0 0 1 0-1.06Z"
            clipRule="evenodd"
          />
        </svg>
      </button>
      <div className={`dropdownMenu ${styles.dropdownMenu} ${isOpen ? '' : 'hidden'}`}>
        {items.map(({ text, onClick }, index) => (
          <button
            key={index}
            type="button"
            className={styles.dropdownItem}
            onClick={() => {
              onClick?.();
              setOpen(false);
            }}
          >
            {text}
          </button>
        ))}
      </div>
    </div>
  );
};

export default DropDown;
