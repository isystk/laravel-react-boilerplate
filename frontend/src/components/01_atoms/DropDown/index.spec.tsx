import React from 'react'
import { render } from '@testing-library/react'
import DropDown, { DropDownProps } from './index'

describe('LandingPageTemplate', () => {
  it('Match Snapshot', () => {
    const props: DropDownProps = {
      label: 'サンプル太郎',
      items: [
        { label: 'リンク1', link: 'http://link1' },
        { label: 'リンク2', link: 'http://link2' },
        { label: 'リンク3', link: 'http://link3' },
      ],
    }
    const { asFragment } = render(<DropDown {...props} />)
    expect(asFragment()).toMatchSnapshot()
  })
})
