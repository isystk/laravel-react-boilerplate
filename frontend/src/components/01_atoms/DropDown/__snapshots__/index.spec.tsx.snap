// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LandingPageTemplate Match Snapshot 1`] = `
<DocumentFragment>
  <div
    class="relative"
  >
    <button
      class="css-0 flex items-center justify-between md:py-2 pl-3 md:pr-4 rounded md:border-0 md:p-0 "
      data-dropdown-toggle="dropdownNavbar"
      id="dropdownNavbarLink"
    >
      <span
        class="w-32 md:w-48 md:pr-4 truncate"
      >
        サンプル太郎
      </span>
      <svg
        aria-hidden="true"
        class="w-5 h-5 ml-1"
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M5.293 7.293a1 1 0 011.414 0L10 10.586l3.293-3.293a1 1 0 111.414 1.414l-4 4a1 1 0 01-1.414 0l-4-4a1 1 0 010-1.414z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
    <div
      class="hidden z-10 font-normal bg-white divide-y divide-gray-100 rounded-lg shadow absolute right-0 w-48"
      id="dropdownNavbar"
    >
      <ul
        aria-labelledby="dropdownLargeButton"
        class="py-2 text-sm text-gray-700 dark:text-gray-400"
      >
        <li>
          <a
            class="block px-4 py-2 hover:bg-gray-100 dark:hover:bg-gray-600 dark:hover:text-white"
            href="#"
          >
            リンク1
          </a>
        </li>
        <li>
          <a
            class="block px-4 py-2 hover:bg-gray-100 dark:hover:bg-gray-600 dark:hover:text-white"
            href="#"
          >
            リンク2
          </a>
        </li>
        <li>
          <a
            class="block px-4 py-2 hover:bg-gray-100 dark:hover:bg-gray-600 dark:hover:text-white"
            href="#"
          >
            リンク3
          </a>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
